File: DONGUK/프로그래머스_WEEK 8/교점에 별 만들기.py
def solution(line):
    answer = []
    
    candidate = []
    for i in range(len(line)):
        a1, b1, c1 = line[i][0], line[i][1], line[i][2]
        for j in range(i+1, len(line)):
            
            a2, b2, c2 = line[j][0], line[j][1], line[j][2]
            
            x, y = 0, 0
            if (a1*b2) - (b1*a2) == 0:
                continue
            x = (b1*c2-c1*b2) / (a1*b2 - b1*a2)
            y = (c1*a2-a1*c2) / (a1*b2 - b1*a2)
            
            if int(x) == x and int(y) == y:
                if (x, y) not in candidate:
                    candidate.append((int(x), int(y)))
    
                                     
    candidate.sort(key = lambda x : x[0])
    minx, maxx = candidate[0][0], candidate[-1][0]
    
    candidate.sort(key = lambda x : x[1])
    miny, maxy = candidate[0][1], candidate[-1][1]
    
    
    
    arr = [["." for j in range(maxx - minx+1)] for i in range(maxy - miny+1)]
    
    for x, y in candidate: 
        r, c = y - miny, x - minx
        
        arr[r][c] = "*"
        
    return ["".join(item) for item in arr][::-1]

File: JINSEOP/프로그래머스_WEEK8/교점에 별 만들기.cpp
#include <string>
#include <vector>
#include <climits>
using namespace std;

long long max_x = LLONG_MIN, max_y = LLONG_MIN, min_x = LLONG_MAX, min_y = LLONG_MAX;
vector<pair<long long, long long>> intersections;
vector<string> res;
void get_intersection(vector<vector<int>> line) {
    for (int i = 0; i < line.size(); i++) {
        for (int j = i+1; j < line.size(); j++) {
            long long x_numerator = (line[i][1] * line[j][2] * 1LL) - (line[i][2] * line[j][1] * 1LL);
            long long y_numerator = (line[i][2] * line[j][0] * 1LL) - (line[i][0] * line[j][2] * 1LL);
            long long denominator = (line[i][0] * line[j][1] * 1LL) - (line[i][1] * line[j][0] * 1LL);
            
            if (denominator == 0) continue;
            if (((x_numerator % denominator) != 0) || ((y_numerator % denominator) != 0))
                continue;

            long long x = x_numerator / denominator;
            long long y = y_numerator / denominator;
            intersections.push_back({ x, y });

            max_x = max(max_x, x); max_y = max(max_y, y);
            min_x = min(min_x, x); min_y = min(min_y, y);
        }
    }
}

vector<string> solution(vector<vector<int>> line) {

    get_intersection(line);

    string temp = string(max_x - min_x + 1, '.');
    vector<string> answer(max_y - min_y + 1, temp);

    for (auto intersection : intersections)
        answer[abs(intersection.second - max_y)][abs(intersection.first - min_x)] = '*';

    return answer;
}

// Long Long type 다루는 법을 생각해보자.
